src/stream/src/executor/asof_join.rs

```diff
@@ -0,0 +1,1379 @@
+            chunk_size,
+            actual_output_data_types.to_vec(),
+            side_update.i2o_mapping.clone(),
+            side_match.i2o_mapping.clone(),
+        );
+
+        let join_matched_rows_metrics = ctx
+            .streaming_metrics
+            .join_matched_join_keys
+            .with_guarded_label_values(&[
+                &ctx.id.to_string(),
+                &ctx.fragment_id.to_string(),
+                &side_update.ht.table_id().to_string(),
+            ]);
+
+        let keys = K::build_many(&side_update.join_key_indices, chunk.data_chunk());
+        for (r, key) in chunk.rows_with_holes().zip_eq_debug(keys.iter()) {
+            let Some((op, row)) = r else {
+                continue;
+            };
+            let mut join_matched_rows_cnt = 0;
+
+            Self::evict_cache(side_update, side_match, cnt_rows_received);
+
+            let matched_rows = if !side_update.ht.check_inequal_key_null(&row) {
+                Self::hash_eq_match(key, &mut side_match.ht).await?
+            } else {
+                None
+            };
+            let inequal_key = side_update.ht.serialize_inequal_key_from_row(row);
+
+            if let Some(matched_rows) = matched_rows {
+                let update_rows = Self::hash_eq_match(key, &mut side_update.ht).await?.expect("None is not expected because we have checked null in key when getting matched_rows");
+                let right_inequality_index = update_rows.inequality_index();
+                let (row_to_delete_r, row_to_insert_r) =
+                    if let Some(pks) = right_inequality_index.get(&inequal_key) {
+                        assert!(!pks.is_empty());
+                        let row_pk = side_match.ht.serialize_pk_from_row(row);
+                        match op {
+                            Op::Insert | Op::UpdateInsert => {
+                                // If there are multiple rows match the inequality key in the right table, we use one with smallest pk.
+                                let smallest_pk = pks.first_key_sorted().unwrap();
+                                if smallest_pk > &row_pk {
+                                    // smallest_pk is in the cache index, so it must exist in the cache.
+                                    if let Some(to_delete_row) = update_rows
+                                        .get_by_indexed_pk(smallest_pk, &side_update.all_data_types)
+                                    {
+                                        (
+                                            Some(Either::Left(to_delete_row?.row)),
+                                            Some(Either::Right(row)),
+                                        )
+                                    } else {
+                                        // Something wrong happened. Ignore this row in non strict consistency mode.
+                                        (None, None)
+                                    }
```

Comment:
need to check with `enable_strict_consistency()`

Reply:
Checked inside the `get_by_indexed_pk`